### DISCUSSION TITLE
SMR-CZ-040W-WIF-16A-POW-US - Have 12 on Networks, but latest additions not seen

### DISCUSSION THREAD

**Discussion by MarkCupitt** [jasonacox/tinytuya#510] [General]
Hi All, I have a bunch of Switches SMR-CZ-040W-WIF-16A-POW-US on my network. Up until recently, I have been seeing them in  tiny-tuya.

I started adding additional switches last couple of days, SmartLife sees and controls them happily, they are in iot.tuya.com quite happily, they are downloaded in devices.json, I see them with an Ip Address in my DHCP Server quite happily, but the tinytuya scan does not see the newer Ones

I have tried using a 120 second timeout 

It seems that all newer ones are not seen, and a couple of older ones are not seen by tinytuya 

I double checked taht all my devices were on the same subnet as well 

Here are two screenshots of the same model, one that is seen and one that is not

Its almost like there is a limit on the number  of a specific model that can be scanned, but I cannot see anything like that in the code

Any thoughts, happy to debug, but not quite sure where to start

![image](https://github.com/jasonacox/tinytuya/assets/5503184/e1f8b2e3-f444-4957-af8c-6ecfbd91b736)

![image](https://github.com/jasonacox/tinytuya/assets/5503184/f935d0a3-82ec-422d-90c4-3c41f872aa34)

Many thanks

**Comment by MarkCupitt** [comment]
This is the JSON from devices.json for the device that is not seen by the scan

```
    {
        "name": "Fridge Main",
        "id": "ebcc2d8e4d9b52cc22eeq1",
        "key": "y7omD3+e09m.fBx2",
        "mac": "c4:82:e1:4f:80:2a",
        "uuid": "f623f3641a809b39",
        "category": "cz",
        "product_name": "smart plug",
        "product_id": "uqehhcrmk5depvtl",
        "biz_type": 18,
        "model": "\u767e\u89c6\u76fe\u8ba1\u91cf",
        "sub": false,
        "icon": "https://images.tuyaus.com/smart/icon/ay15107356719526e3ps/f1d0c8ad0822636028bd9117522687db.png",
        "mapping": {
            "1": {
                "code": "switch_1",
                "type": "Boolean",
                "values": {}
            },
            "9": {
                "code": "countdown_1",
                "type": "Integer",
                "values": {
                    "unit": "s",
                    "min": 0,
                    "max": 86400,
                    "scale": 0,
                    "step": 1
                }
            },
            "17": {
                "code": "add_ele",
                "type": "Integer",
                "values": {
                    "unit": "kwh",
                    "min": 0,
                    "max": 50000,
                    "scale": 3,
                    "step": 100
                }
            },
            "18": {
                "code": "cur_current",
                "type": "Integer",
                "values": {
                    "unit": "mA",
                    "min": 0,
                    "max": 80000,
                    "scale": 0,
                    "step": 1
                }
            },
            "19": {
                "code": "cur_power",
                "type": "Integer",
                "values": {
                    "unit": "W",
                    "min": 0,
                    "max": 200000,
                    "scale": 1,
                    "step": 1
                }
            },
            "20": {
                "code": "cur_voltage",
                "type": "Integer",
                "values": {
                    "unit": "V",
                    "min": 0,
                    "max": 5000,
                    "scale": 1,
                    "step": 1
                }
            },
            "38": {
                "code": "relay_status",
                "type": "Enum",
                "values": {
                    "range": [
                        "power_off",
                        "power_on",
                        "last"
                    ]
                }
            },
            "39": {
                "code": "light_mode",
                "type": "Enum",
                "values": {
                    "range": [
                        "relay",
                        "pos",
                        "none"
                    ]
                }
            },
            "40": {
                "code": "child_lock",
                "type": "Boolean",
                "values": {}
            },
            "41": {
                "code": "cycle_time",
                "type": "String",
                "values": "{}"
            },
            "42": {
                "code": "random_time",
                "type": "String",
                "values": "{}"
            },
            "43": {
                "code": "switch_inching",
                "type": "String",
                "values": "{}"
            },
            "51": {
                "code": "overcharge_switch",
                "type": "Boolean",
                "values": {}
            }
        },
        "ip": "",
        "version": ""
    },
    ```
    
    This is the JSON for the same model Plug, bought together that IS seen in the scan
    
    ```
        {
        "name": "Fridge Backup",
        "id": "eb462657f66ad83c67avat",
        "key": "^uxEG^oFsKI-j/a'",
        "mac": "c4:82:e1:96:43:7d",
        "uuid": "732c330c2d9a647c",
        "category": "cz",
        "product_name": "Smart plug",
        "product_id": "o7ve1lp3gfjvsvyo",
        "biz_type": 18,
        "model": "BP20V4",
        "sub": false,
        "icon": "https://images.tuyaus.com/smart/icon/non-session-user/1655258627f3c9dadb5e2.png",
        "mapping": {
            "1": {
                "code": "switch_1",
                "type": "Boolean",
                "values": {}
            },
            "9": {
                "code": "countdown_1",
                "type": "Integer",
                "values": {
                    "unit": "s",
                    "min": 0,
                    "max": 86400,
                    "scale": 0,
                    "step": 1
                }
            },
            "17": {
                "code": "add_ele",
                "type": "Integer",
                "values": {
                    "min": 0,
                    "max": 50000,
                    "scale": 3,
                    "step": 100
                }
            },
            "18": {
                "code": "cur_current",
                "type": "Integer",
                "values": {
                    "unit": "mA",
                    "min": 0,
                    "max": 30000,
                    "scale": 0,
                    "step": 1
                }
            },
            "19": {
                "code": "cur_power",
                "type": "Integer",
                "values": {
                    "unit": "W",
                    "min": 0,
                    "max": 80000,
                    "scale": 1,
                    "step": 1
                }
            },
            "20": {
                "code": "cur_voltage",
                "type": "Integer",
                "values": {
                    "unit": "V",
                    "min": 0,
                    "max": 5000,
                    "scale": 1,
                    "step": 1
                }
            },
            "38": {
                "code": "relay_status",
                "type": "Enum",
                "values": {
                    "range": [
                        "power_off",
                        "power_on",
                        "last"
                    ]
                }
            },
            "39": {
                "code": "overcharge_switch",
                "type": "Boolean",
                "values": {}
            },
            "40": {
                "code": "light_mode",
                "type": "Enum",
                "values": {
                    "range": [
                        "relay",
                        "pos",
                        "none"
                    ]
                }
            },
            "41": {
                "code": "child_lock",
                "type": "Boolean",
                "values": {}
            },
            "42": {
                "code": "cycle_time",
                "type": "String",
                "values": "{}"
            },
            "43": {
                "code": "random_time",
                "type": "String",
                "values": "{}"
            },
            "44": {
                "code": "switch_inching",
                "type": "String",
                "values": "{}"
            }
        },
        "ip": "192.168.2.182",
        "version": "3.4"
    },
    ```
    
    I checked My DHCP Server (using the MAC Addresses)
    
   Fridge Main has an IP address of 192.168.2.173
   Fridge Backup has an Ip address of  192.168.2.182

**Comment by uzlonewolf** [comment]
Hi @MarkCupitt,

First, have you made sure the Smart Life app and any other Tuya implementations (LocalTuya, HomeAssistant, etc) are closed?  A lot of devices stop broadcasting once a connection is established.

What version of TinyTuya do you have?  `python3 -m tinytuya -h` will tell you.

Have you tried running a force scan?  `python3 -m tinytuya scan -f`

If those don't work then does a scan with debug enabled show anything?  `python3 -m tinytuya scan -d`

**Reply by MarkCupitt** [reply]
@uzlonewolf .. Yes to all except the debug, here are the results

Sent error output to a file then searched the file for the device id `ebcc2d8e4d9b52cc22eeq1` (Which is the Fridge Main in devices.json) and that string is not shown in the output at all

Coudl it be that tinytuya is ignoring it form the devices.json??

I notice that it does not have a "version" attribute in the json which is a little weird as others that work do

**Reply by MarkCupitt** [reply]
@uzlonewolf I also did the same process with the -force option, that device id does not appear in the output anywhere

**Reply by uzlonewolf** [reply]
~~So, what version of TinyTuya do you have?~~ Nvm, I see in the screenshot it's 1.14.0

Does the debug output say anything about "Unexpected payload from ...", "Invalid UDP Packet from ...", or being unable to decode something?

**Reply by MarkCupitt** [reply]
Ill check that tomorrow, thanks

**Comment by MarkCupitt** [comment]
@uzlonewolf I noticed this in tinytuya scanner.,py, shodul I uncomment?

```
# Optional libraries required for forced scanning
#try:
#    from getmac import get_mac_address
#    SCANLIBS = True
#except:
#    SCANLIBS = False
```

**Reply by MarkCupitt** [reply]
Did a `pip install getmac` and uncommented, then did a force scan again, same result, taht device id is not in teh output

**Reply by uzlonewolf** [reply]
That was from an attempt at finding the MAC address for discovered devices.  It was removed because it was extremely slow and didn't really work well with some of the new options such as force scanning.  I should probably delete it completely...

**Comment by MarkCupitt** [comment]
@uzlonewolf 

Scanner.py, around line 182, Im wondering if teh empty version is forcing a "3.1" an dit needs a 3.5

```
        if not self.deviceinfo['version']:
            self.deviceinfo['version']  = 3.1
```

**Reply by MarkCupitt** [reply]
Tried forcing the versions .. 3.1, 3.2, 3.3, 3.4, also added in the actual Ip address in devices.json from DHCP, like this, 

```
        "ip": "192.168.2.173",
        "version": "3.4"
```

No Joy

**Reply by uzlonewolf** [reply]
The scanner does not read "ip" or "version" from devices.json, it tries to detect both via device broadcast packets and/or a force scan.

Since you already know the IP and version, does the monitor example https://github.com/jasonacox/tinytuya/blob/master/examples/monitor.py work?

**Reply by MarkCupitt** [reply]
Ill try this tomorrow .. Thanks

**Comment by MarkCupitt** [comment]
@uzlonewolf I have two Public Ip addresses, (2 Starlinks on a LB Router), I see the Fridge backup is on a 14.1.64.* address when I check teh network on teh app, nut the Fridge Main which tinytuya cannot see is on a 65.* network

Im not sure how the UDP listening is configured, but there would be different gateways I guess

COudl this be related??

**Reply by uzlonewolf** [reply]
As long as both devices are on the same Layer 2 network it should just work.  Local connections do not care about what the WAN address is or even if it has WAN connectivity at all.

**Reply by MarkCupitt** [reply]
Yep, I understand that, but Im wondering if the microtik router/gateway is trapping the broadcast packets for one specific outbound IP and not forwarding them ..  .. Ill investigate with wireshark

**Reply by uzlonewolf** [reply]
Broadcasts are a Layer 2 thing, they do not go through routers/gateways and are never forwarded.

Wireshark is a good idea and will tell you if the devices are even broadcasting.  They're extremely rare but some devices do not allow local connections at all and do not broadcast.  Most devices which do broadcast and allow connections stop broadcasting once a connection is made.  Rebooting the device while Wireshark is running is also something to try.

**Reply by MarkCupitt** [reply]
The Microtik is actually a router/gateway/switch .. It seems that any device with one specific  public IP address in devices.json is not being seen which is really weird. I'll try your suggestion in the examples to prove whether its the device or not, or network related. The Tuya devices are identical, all bought at the same time, no obvious differences between them other than the timeline on setup ..

A real mystery ..

**Reply by uzlonewolf** [reply]
They may have been purchased at the same time and might even look the same on the outside, but they are different devices as shown by the devices.json data:
```
        "product_name": "smart plug",
        "product_id": "uqehhcrmk5depvtl",
        "model": "\u767e\u89c6\u76fe\u8ba1\u91cf",
        "icon": "https://images.tuyaus.com/smart/icon/ay15107356719526e3ps/f1d0c8ad0822636028bd9117522687db.png",
...
        "product_name": "Smart plug",
        "product_id": "o7ve1lp3gfjvsvyo",
        "model": "BP20V4",
        "icon": "https://images.tuyaus.com/smart/icon/non-session-user/1655258627f3c9dadb5e2.png",
```

**Reply by MarkCupitt** [reply]
Ahh, I missed that, thanks for pointing that out .. I'll check the rest and
see if thst is consistent

**Reply by MarkCupitt** [reply]
INteresting, so the unicode for `\u767e\u89c6\u76fe\u8ba1\u91cf` converts to "百视盾计量" which Google Translate says is `Besight Shield Metering`

**Reply by MarkCupitt** [reply]
So I have 5 devices with that Unicode Model, 3 are NOT Seen by tinytuya, 2 are seen .. All 5 devices have the same Product Id: `uqehhcrmk5depvtl`

**Reply by MarkCupitt** [reply]
SO I cannot see ANY of those 5 devices (searching by product_id)  and Name, in a debug log, normal scan log, or force log, However 2 of the 5 ARE seen and showing on the Web Page 

attaching logs and devices.json  to see if anyone can see if I missed something obvious ...

The three Device Ids that DO NOT show up are
`ebb4a8d6f4f20dd7e6osmb`
`eba0f51072f1a7384da9sz`
`eb52c2498cf7af8fdeq7p2`

[devices.json](https://github.com/user-attachments/files/15942653/devices.json)

[scan.log](https://github.com/user-attachments/files/15942621/scan.log)

[force.log](https://github.com/user-attachments/files/15942620/force.log)

[debug.log](https://github.com/user-attachments/files/15942619/debug.log)

**Reply by uzlonewolf** [reply]
Hmm, I'm not seeing any unknown packets in those logs, so it looks like no broadcast packets are being received from those devices.  I'm leaning towards a network issue here.  Can you ping them?  If so, does the monitor script from https://github.com/jasonacox/tinytuya/blob/master/examples/monitor.py work with them?

**Reply by MarkCupitt** [reply]
SOme Further Information, 

I set up a Wireshark Monitor using the Mac Address, for one of teh devices that was not seen by tinytuya

I powered the device off, waited 60 seconds and powered it on I can see packets  when I use Smart Life App to control the device .. Expected

Now, when i close the Smart Lfe App, and power off and on agfain, I see the folowinfg, what is interesting, it is stil announcing to Ip 192.168.2.106, which is my Mobile Phone with te hSMart LKiufe app on it. Running A force, Debugh and normal scan deos NOT generate any additional packets, The only packets I see are as per screenshot below .. any thoughts 

![image](https://github.com/jasonacox/tinytuya/assets/5503184/ad9a1277-4d8b-4be2-b6ff-3552e807d2a2)

**Reply by uzlonewolf** [reply]
While looking into this I realized the latest TinyTuya update broke the force scan, so a force scan isn't going to work.

The 2 things I'd try at this point are completely disconnecting the phone from the network (the Smart Life app likes to run in the background) and rebooting the device, and the monitor script I posted above.

**Reply by MarkCupitt** [reply]
ANother thing I noticed is that of all teh same devices I purchased, soe are showing in devices.json as Version 3.4, amd sopm,e are showing as version "", however the ones taht tinytuya sees, seems ot be detecting V 3.5

**Reply by MarkCupitt** [reply]
Ok, cool, Ill try that now @uzlonewolf thanks on the force scan .. I did not realize that .. good info

**Reply by uzlonewolf** [reply]
If you want a fixed force-scan you can check out https://github.com/jasonacox/tinytuya/pull/511

**Reply by MarkCupitt** [reply]
@uzlonewolf  OK, I ran monitor.py and was able to get a status from a device that does not show in the web app as On Line, here is the config I used ..

`d = tinytuya.OutletDevice('eb52c2498cf7af8fdeq7p2', '192.168.2.106', '-4[<i7I}B+3.jO:4', version=3.5, persist=True)`

V3.4 returned an  error to check the device key and version

Given the number of non alpha chars in the device key, could that be the issue, tinytyua does not deal with them, or one of them, maybe the colon is upsetting it??

I don't quite understand it all, I presume the scan just looks for packets on the way out, but the force will try and actively ping the device direct, if force is not working, then maybe why we did not see the devices..

I'm still a little confused, given that the device is emitting a status when asked so why is it not being seen? Version issue??

**Reply by MarkCupitt** [reply]
@uzlonewolf I picked up your forcescan fixes, and gave it a try, and no change on my issue, HOWEVER ... I noticed this

`Versions: 3.1: 1, 3.3: 8, 3.4: 22`  in the output

I know for a fact that the devices with issues are V 3.5 (there are 5 of them) and it seems to be either not trying that version, or getting some error that is not reported in the logs

The Web Interface reports `Versions: 3.1: 1, 3.3: 8, 3.4: 22`  Force Scan reports `Scan Complete!  Found 31 devices.`, the difference is, TaDa, 5 devices short in the force scan

Log FIle
[force.log](https://github.com/user-attachments/files/15943310/force.log)

**Comment by MarkCupitt** [comment]
@jasonacox @uzlonewolf Yep, I actually had disabled wifi for my phone with Smartlife on it, and restarted all the devices.

The pull request does address the 3.5 issue which was the root cause of my issue, and I accept that the devices are "different" and do something different with Broadcast packets.

Ill adjust my use case to work with this approach as I will know the IP addresses and can access them directly. Im comfortable with that, so I would say the PR is Ok

Thanks for the prompt fixes and responses, they are very much appreciated

**Reply by jasonacox** [reply]
Thanks, I'm going to merge and publish.  Thanks for raising this @MarkCupitt and for your tests and collaboration!

**Comment by MarkCupitt** [comment]
Just FYI, my approach will be something like the following:

Add New Device
Make DHCP allocated device IP static, so it is always known, and have a map of devices to IPs somewhere
Use TinyTuya Wizard to scan and get active devices, continually, so snapshot.json is always current
Any devices showing offline, ie: is not in snapshot.json (or grab list via an API call to tinytuya) use something like monitor.py to actively poll the device to determine if it is on line

There may be a need to access the Tuya API directly for some devices, like sensors that only emit occasionally, I am still looking into this 

I will probably need another map file to keep this all together

Once I am able to access all the known devices, I will poll them on specific intervals, with the interval determined by a value in the map file to get status and publish to MQTT for consumption in my Farm Management System

I am doing this because I have Weather, Solar Inverter, and Battery Systems I also monitor via MQTT and want to build an integrated management system. I was using HA, but I have not found it to be too unreliable for my needs

Ill likely share what I have done via PR's, Ill try an bundle it as a separate service, if any interest you will be welcome to include it in tinytuya

**Comment by uzlonewolf** [comment]
Just curious, if you run this in one terminal:
```python
import tinytuya
import socket
import time

bcast = b'{"from":"app","ip":"192.168.2.1"}'


bcast_port = 7000
bcast_to = '255.255.255.255'
bcast_idx = 0
bcast_cmd = 0x25
bcast_retcode = None

bsock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) # UDP
bsock.setsockopt(socket.SOL_SOCKET, socket.SO_BROADCAST, 1)

while True:
    bcast_msg = tinytuya.TuyaMessage( bcast_idx, bcast_cmd, bcast_retcode, bcast, 0, True, tinytuya.PREFIX_6699_VALUE, True )
    bcast_data = tinytuya.pack_message( bcast_msg, hmac_key=tinytuya.udpkey )
    print("broadcast encrypted=%r" % bcast_data.hex() )
    bsock.sendto( bcast_data, (bcast_to, bcast_port) )
    time.sleep(5)
```
Does `python3 -m tinytuya scan` in another then pick up those missing devices?
